src/einspline/bspline_eval_d_std.cpp

```diff
@@ -28,43 +29,129 @@ extern const double* restrict   Ad;
 extern const double* restrict  dAd;
 extern const double* restrict d2Ad;
 
+/**
+ * As shipped einspline does no checking of bounds of indexes calculated from
+ *  the input.
+ */
+void check_ipart(const int ipart,
+		 const Ugrid* const restrict grid,
+		 const BCtype_d* const restrict BC)
+{
+#ifndef NDEBUG
+  assert( ipart >= 0 );
+  if (BC->lCode == NATURAL)
+  {
+    assert( ipart <= grid->num -2 );
+  }
+  else
+  {
+    assert( ipart <= grid->num -1 );
+  }
+#endif
+}
+
 /************************************************************/
 /* 1D double-precision, real evaluation functions           */
 /************************************************************/
 
-/* Value only */
-void
-eval_UBspline_1d_d (UBspline_1d_d * restrict spline,
-                    double x, double* restrict val)
+/** Offset is i remainder is t
+ *  einspline as shipped just repeats this code over and over and over
+ */
+
+void check_coeff_offset_UBspline_1d_d(const UBspline_1d_d * const restrict spline,
+					     double x,
+					     int& i,
+				      double& t)
 {
   x -= spline->x_grid.start;
   double u = x*spline->x_grid.delta_inv;
-  double ipart, t;
+  double ipart;
   t = modf (u, &ipart);
-  int i = (int) ipart;
+  i = ipart;
+  check_ipart(ipart, &(spline->x_grid), &(spline->xBC));
+}
+
+void check_coeff_offset_UBspline_2d_d(const UBspline_2d_d * const restrict spline,
```

Comment:
Can we just call the 1D twice or triple times instead of making a separate function for each case?

Reply:
I believe not because of the x_grid y_grid z_grid stuff.

Reply:
this is a nice illustration of why using names instead of template parameters is a bad idea

Reply:
Passing arguments?

Reply:
A simple operation replicated 6 times is not a good thing.

Reply:
previously replicated in every eval I didn't invent this library.

Reply:
I'm not saying the old way is good or bad. It was simple and fine. Now there are more stuff operated on each direction. It is worth adding a function but not one for each dimension case with contents duplicated. I prefer one function and called for each direction in the eval function.

Reply:
feel free